### PR TITLE
Add auto-match in assisted tagging

### PR DESCRIPTION
**Pull request by AdityaKhatri**
## Changes

* Add auto match functionality in assisted tagging to match titles directly

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] codegen errors
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [x] permission checks
- [x] translations

### CODE REVIEW

File: app/views/AnalyticalFramework/AssistedTagging/index.tsx [hunk 1]
```diff
@@ -212,6 +226,30 @@ function AssistedTagging<K extends string>(props: Props<K>) {
...
+    const handleAutoMatchClick = useCallback(() => {
+        const items = widgets?.map((widget) => {
+            if (widget.widgetId === 'MATRIX2D') {
+                return getMatrix2dThings(widget);
+            }
+            return [];
+        }).flat();
+        const mapping = items?.map((item) => {
+            const tagId = predictionTagsMapByName?.[item?.label?.toLowerCase()];
+            console.warn('here', item.label, predictionTagsMapByName);
```
Comment (samshara) [inline app/views/AnalyticalFramework/AssistedTagging/index.tsx]: remove console.warn() here.

File: app/views/AnalyticalFramework/AssistedTagging/utils.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                },
+                clientId: randomString(),
+            }))
+        )).flat() ?? [];
+
+    return [
+        ...columns,
+        ...subRows,
+        ...subColumns,
+    ];
```
Comment (samshara) [inline app/views/AnalyticalFramework/AssistedTagging/utils.ts]: any reason why we aren't mapping `rows` here?

**Review by samshara — APPROVED** [review]

File: app/views/AnalyticalFramework/AssistedTagging/index.tsx [hunk 3]
```diff
@@ -243,14 +300,25 @@ function AssistedTagging<K extends string>(props: Props<K>) {
...
+                            value={assistedTaggingEnabled}
+                            onChange={onAssistedTaggingStatusChange}
+                            disabled={pending}
+                            label="Active"
+                        />
+                        <ConfirmButton
+                            name={undefined}
+                            onConfirm={handleAutoMatchClick}
+                            message="Auto-matching will remove all the current mappings and replace them with the recommended ones. Are you sure you want to auto-match?"
+                            variant="tertiary"
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/index.tsx:318]: Don't we need to disable this when `pending`?

File: app/views/AnalyticalFramework/AssistedTagging/index.tsx [hunk 4]
```diff
@@ -212,6 +229,46 @@ function AssistedTagging<K extends string>(props: Props<K>) {
...
+            }
+            if (widget.widgetId === 'ORGANIGRAM') {
+                return getOrganigramPossibleMappings(widget);
+            }
+            return [];
+        }).flat();
+
+        const mapping = items?.map((item) => {
+            const tagId = predictionTagsMapByName?.[item?.label?.toLowerCase()];
+            if (tagId) {
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/index.tsx]: When will we not have tagId? When af changes?

File: app/views/AnalyticalFramework/AssistedTagging/utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                        label: 'Sub row 2 3',
+                        order: 3,
+                    },
+                ],
+            },
+        ],
+    },
+};
+
+const matrix2dPossibleMapping = [
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/utils.test.ts]: Let's define the type for possible mapping

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/AnalyticalFramework/AssistedTagging/utils.test.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        },
+    },
+];
+
+test('get matrix possible mappings', () => {
+    expect(getMatrix2dPossibleMappings(undefined)).toStrictEqual([]);
+    expect(
+        getMatrix2dPossibleMappings(matrix2dSample),
+    ).toStrictEqual(matrix2dPossibleMapping);
+});
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/utils.test.ts:419]: Let's add tests for other widgets as well.

**Review by tnagorra — APPROVED** [review]

File: app/views/AnalyticalFramework/AssistedTagging/index.tsx [hunk 7]
```diff
@@ -212,6 +224,51 @@ function AssistedTagging<K extends string>(props: Props<K>) {
...
+                            ...item,
+                            tag: tag.id,
+                            clientId: randomString(),
+                        },
+                        ...(acc ?? []),
+                    ]);
+                }
+                return acc;
+            }, undefined as Omit<MappingsItem, 'id'>[] | undefined)
+        )).filter(isDefined).flat();
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/index.tsx]: Pikachu!!

File: app/views/AnalyticalFramework/AssistedTagging/index.tsx (hunk 3 above)
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/index.tsx:318]: Bulbasaur!

File: app/views/AnalyticalFramework/AssistedTagging/utils.test.ts [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+        rows: [
+            {
+                key: 'row-1',
+                label: 'row-1',
+                order: 1,
+                color: '',
+                cells: [
+                    {
+                        key: 'sub-row-1-1',
+                        label: 'sub-row-1-1',
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/utils.test.ts]: Charizard!

File: app/views/AnalyticalFramework/AssistedTagging/utils.test.ts [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+    id: '12',
+    clientId: 'sample-scale',
+    widgetId: 'SCALE',
+    properties: {
+        options: [
+            {
+                key: 'option-1',
+                label: 'option-1',
+                order: 1,
+                color: '',
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/utils.test.ts]: Dragonite!

File: app/views/AnalyticalFramework/AssistedTagging/utils.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    ScaleMappingsItem,
+    SelectMappingsItem,
+    MultiSelectMappingsItem,
+    OrganigramWidget,
+    OrganigramMappingsItem,
+} from '#types/newAnalyticalFramework';
+
+import { getOrganigramFlatOptions } from './WidgetTagList/OrganigramTagInput/utils';
+
+type Possible<T> = Omit<T, 'tag' | 'clientId' | 'id'> & { label: string };
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/utils.ts]: Squirtle!

File: app/views/AnalyticalFramework/AssistedTagging/utils.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                    rowKey: row.key,
+                },
+            }))
+        )).flat() ?? [];
+
+    return [
+        ...columns,
+        ...subColumns,
+        ...subRows,
+    ] as Matrix2dPossibleMapping[];
```
Comment (tnagorra) [inline app/views/AnalyticalFramework/AssistedTagging/utils.ts]: Slowpoke!

**Review by tnagorra — CHANGES_REQUESTED** [review]

**Review by tnagorra — APPROVED** [review]

**Review by subinasr — APPROVED** [review]

**Pull request merged**